internal/pkg/istio/config_factory.go

```diff
@@ -222,97 +228,120 @@ func (cf *ConfigFactory) EnvoyFilters() []*v1alpha3.EnvoyFilter {
 }
 
 func (cf *ConfigFactory) ServiceEntries() ([]*v1alpha3.ServiceEntry, error) {
-	remote := cf.cfg.MeshPeers.Remote
 
-	if len(remote.Addresses) == 0 {
-		return nil, nil
-	}
+	var serviceEntries []*v1alpha3.ServiceEntry
+	serviceEntriesByHost := make(map[string]*v1alpha3.ServiceEntry, len(cf.cfg.MeshPeers.Remotes))
 
-	var resolution istionetv1alpha3.ServiceEntry_Resolution
-	if networking.IsIP(cf.cfg.MeshPeers.Remote.Addresses[0]) {
-		resolution = istionetv1alpha3.ServiceEntry_STATIC
-	} else {
-		resolution = istionetv1alpha3.ServiceEntry_DNS
-	}
+	for _, remote := range cf.cfg.MeshPeers.Remotes {
+		if len(remote.Addresses) == 0 {
+			continue
+		}
 
-	serviceEntries := []*v1alpha3.ServiceEntry{cf.serviceEntryForRemoteFederationController()}
-	for _, importedSvc := range cf.importedServiceStore.From(remote) {
-		_, err := cf.serviceLister.Services(importedSvc.Namespace).Get(importedSvc.Name)
-		if err != nil {
-			if !errors.IsNotFound(err) {
-				return nil, fmt.Errorf("failed to get Service %s/%s: %w", importedSvc.Name, importedSvc.Namespace, err)
-			}
-			// Service doesn't exist - create ServiceEntry.
-			var ports []*istionetv1alpha3.ServicePort
-			for _, port := range importedSvc.Ports {
-				ports = append(ports, &istionetv1alpha3.ServicePort{
-					Name:       port.Name,
-					Number:     port.Number,
-					Protocol:   port.Protocol,
-					TargetPort: port.TargetPort,
+		serviceEntries = append(serviceEntries, cf.serviceEntryForRemoteFederationController(remote))
+
+		var resolution istionetv1alpha3.ServiceEntry_Resolution
+		if networking.IsIP(remote.Addresses[0]) {
+			resolution = istionetv1alpha3.ServiceEntry_STATIC
+		} else {
+			resolution = istionetv1alpha3.ServiceEntry_DNS
+		}
+
+		for _, importedSvc := range cf.importedServiceStore.From(remote) {
+			_, err := cf.serviceLister.Services(importedSvc.Namespace).Get(importedSvc.Name)
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					return nil, fmt.Errorf("failed to get Service %s/%s: %w", importedSvc.Name, importedSvc.Namespace, err)
+				}
+				// Service doesn't exist - create ServiceEntry.
+				// TODO(multi-peer) handle naming clash & different resolution strategy (known limitation?)
+				// What if we happen to have multiple remotes with exported serviceName/serviceNamespace but:
+				// - each remote has different address defined (leading to static or DNS resolution)
+				// - services per remote peer differ on service ports
```

Comment:
This is very interesting case, and we can't handle it. It shows how important it is to support custom domains ASAP.